app/src/main/java/ch/epfl/cs311/wanderwave/viewmodel/TrackListViewModel.kt

```diff
@@ -31,43 +32,52 @@ constructor(
     private val appDatabase: AppDatabase,
     private val repository: TrackRepository,
     private val profileRepository: ProfileRepository,
-    private val authenticationController: AuthenticationController
+    private val authenticationController: AuthenticationController,
+    private val recentlyPlayedRepository: RecentlyPlayedRepository
 ) : ViewModel(), SpotifySongsActions {
 
   private val _uiState = MutableStateFlow(UiState(loading = true))
   val uiState: StateFlow<UiState> = _uiState
 
   private var _searchQuery = MutableStateFlow("")
+  private var observeTracksJob: Job? = null
 
   private val _nbrLikedSongs = MutableStateFlow(0)
   override val nbrLikedSongs: StateFlow<Int> = _nbrLikedSongs
 
   fun loadTracksBasedOnSource(index: Int) {
     viewModelScope.launch {
+      observeTracksJob?.let { it.cancel() }
       _uiState.value = _uiState.value.copy(loading = true)
       when (index) {
-        0 ->
-            _uiState.value =
-                _uiState.value.copy(
-                    tracks =
-                        spotifyController.recentlyPlayedTracks
-                            .value) // TODO:modify for the recnetly played tracks
-        1 -> loadRecentlyAddedTracks() // TODO: modify here for the liked tracks
-        2 -> loadRecentlyAddedTracks() // TODO: modify here for the banned tracks
+        0 -> loadRecentlyPlayedTracks()
+        1 -> loadRecentlyAddedTracks() // TODO: modify here for recently added tracks
+        2 -> loadRecentlyAddedTracks() // TODO: modify here for the liked tracks
+        3 -> loadRecentlyAddedTracks() // TODO: modify here for the banned tracks
       }
     }
   }
 
-  fun loadRecentlyAddedTracks() {
-    viewModelScope.launch {
-      val trackRecords =
-          appDatabase.trackRecordDao().getAllRecentlyAddedTracks().firstOrNull() ?: listOf()
-      val trackDetails =
-          trackRecords.mapNotNull {
-            repository.getItem(it.trackId).firstOrNull()?.getOrElse { null }
+  fun loadRecentlyPlayedTracks() {
+    observeTracksJob =
+        viewModelScope.launch {
+          recentlyPlayedRepository.getRecentlyPlayed().collect { tracks ->
+            _uiState.value = _uiState.value.copy(tracks = tracks, loading = false)
           }
-      _uiState.value = _uiState.value.copy(tracks = trackDetails, loading = false)
-    }
+        }
+  }
+
+  fun loadRecentlyAddedTracks() {
+    observeTracksJob =
+        viewModelScope.launch {
+          val trackRecords =
+              appDatabase.trackRecordDao().getAllRecentlyAddedTracks().firstOrNull() ?: listOf()
```

Comment:
I don't know if the repository implement changing flow features, but what I've done usually, is to collect the flow so that the ui state reflect it instead of taking the first one. Especially as sometimes the first flow has an error etc...
You see